src/traverse/index.js

```diff
@@ -193,22 +199,82 @@ module.exports = {
           }
 
           // Per-node handler.
-          if (typeof handler[node.type] === 'function') {
+          let handlerFunc_pre = null;
+          if (handler[node.type] &&
+              typeof handler[node.type] === 'function') {
+            handlerFunc_pre = handler[node.type].bind(handler);
+          }
+          else if (handler[node.type] &&
+                   typeof handler[node.type] === 'object' &&
+                   typeof handler[node.type].pre === 'function') {
+            handlerFunc_pre = handler[node.type].pre.bind(handler);
+          }
+
+          if (handlerFunc_pre) {
             if (options.asNodes) {
-              handler[node.type](node, parent, prop, index);
+              handlerFunc_pre(node, parent, prop, index);
             } else {
               // A path/node can be removed by some previous handler.
               if (!nodePath.isRemoved()) {
-                const handlerResult = handler[node.type](nodePath);
+                const handlerResult = handlerFunc_pre(nodePath);
                 // Explicitly stop traversal.
                 if (handlerResult === false) {
                   return false;
                 }
               }
             }
           }
+        } // Loop over handlers
+
+      }, // pre func
+
+      /**
+       * Handler on node exit.
+       */
+      post(node, parent, prop, index) {
+        if (!node) { // No node? Not sure how this can happen, but without it many tests fail.
+          return;
         }
-      },
+
+        let parentPath;
+        let nodePath;
+
+        if (!options.asNodes) {
+          parentPath = NodePath.getForNode(parent);
+
+          nodePath = NodePath.getForNode(
+            node,
+            parentPath,
+            prop,
+            index
+          );
+        }
```

Comment:
This part is completely repeated, probably worth extracting to a helper method:

```js
let {
  nodePath,
  parentPath,
} = this._getPathsFor(node, parent, prop, index, options);
```

And the:

```js
_getPathsFor(node, parent, prop, index, options) {
  let parentPath;
  let nodePath;

  if (!options.asNodes) {
    parentPath = NodePath.getForNode(parent);

    nodePath = NodePath.getForNode(
      node,
      parentPath,
      prop,
      index
    );
  }

  return {nodePath, parentPath};
}
```